shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCX.scala

```diff
@@ -245,8 +239,10 @@ class UCX(transport: UCXShuffleTransport, executor: BlockManagerId, rapidsConf:
           // else worker.progress returned 0
           if (rapidsConf.shuffleUcxUseWakeup) {
             drainWorker()
-            withResource(new NvtxRange("UCX Sleeping", NvtxColor.PURPLE)) { _ =>
-              worker.waitForEvents()
+            if (workerTasks.isEmpty) {
```

Comment:
There is. I forgot to go back to this and try to either handle it better, so good eye and thanks for raising it.

The `waitForEvents` depends on native code for it to get unblocked. `signal` triggers that native code and so would events from the transport side, therefore we don't want to hold onto java locks here, otherwise we deadlock.

I think making use of an atomic and signal repeatedly while we set it is the answer. I am looking at that now.

Reply:
@jlowe and I discussed with the UCX team and `waitForEvents` is checking for _any_ events before it returns. So there is actually no race here. I am adding a comment to this line.